compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -584,51 +955,128 @@ void PropagateLinalgTransposePass::runOnOperation() {
-        .insert<NamedOpConversion</*OpType=*/linalg::BatchMatmulTransposeBOp,
-                                  /*ReplacementType=*/linalg::BatchMatmulOp,
-                                  /*inputIdx=*/1>>(
-            context, SmallVector<int64_t>{0, 2, 1});
-    sinkingPatterns
-        .insert<NamedOpConversion</*OpType=*/linalg::BatchMatmulTransposeAOp,
-                                  /*ReplacementType=*/linalg::BatchMatmulOp,
-                                  /*inputIdx=*/0>>(
-            context, SmallVector<int64_t>{0, 2, 1});
     sinkingPatterns.insert<SinkTransposeThroughExtractSlice>(context);
     sinkingPatterns.insert<SinkTransposeThroughExpandShape>(context);
-    sinkingPatterns.insert<FuseTransposeWithGenericConsumer>(context);
+    populateNamedOpSinkingPatterns(context, sinkingPatterns);
     sinkingPatterns.add<PropagateTransposeThroughUnaryElementwiseInput>(
         context, /*benefit=*/2);
+    if (failed(
+            applyPatternsAndFoldGreedily(funcOp, std::move(sinkingPatterns)))) {
+      return signalPassFailure();
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- After canonicalizing transpose in place ---\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Propagate transposes upwards, and fuse with any producer generic ops. Also
+  // propagate reshapes upwards to open up more transpose fusion opportunities.
+  if (!testSinkingOnly) {
+    linalg::ControlFusionFn reshapePropagationFn =
+        [&](OpOperand *fusedOperand) {
+          Operation *producer = fusedOperand->get().getDefiningOp();
+          Operation *consumer = fusedOperand->getOwner();
+          if (!IREE::Flow::isNonNullAndOutsideDispatch({producer, consumer})) {
+            return false;
+          }
+
+          // Do not reshape producer linalg op if it has more than one user.
+          auto producerLinalgOp = dyn_cast<linalg::LinalgOp>(producer);
+          if (!producerLinalgOp || !producerLinalgOp->hasOneUse()) {
+            return false;
+          }
+          // Only reshape generic ops, or any op if aggressive propagation is
+          // enabled.
+          if (!enableAggressivePropagation &&
+              !isa<linalg::GenericOp>(producerLinalgOp)) {
+            return false;
+          }
+          // Only propagate expand_shape ops up through producers because it
+          // is always possible to bubble a transpose through an collapse_shape
+          // and thus is handled separately.
+          if (!isa<tensor::ExpandShapeOp>(consumer)) {
+            return false;
+          }
+          // Only propagate if the immediate consumer of the reshape is a
+          // transpose.
+          return consumer->hasOneUse() &&
+                 llvm::isa<linalg::TransposeOp>(*(consumer->user_begin()));
+        };
+    RewritePatternSet bubblingPatterns(context);
+    linalg::populateFoldReshapeOpsByExpansionPatterns(bubblingPatterns,
```

Comment:
I was under the impression that the reshape propagation thing would be changed. Is that not pushed yet?

Reply:
yes, wasn't pushed yet. I dropped the collapsing patterns and only kept the expansion patterns.